### ISSUE TITLE
Rework incorrect indentation recovery

### ISSUE THREAD

**Issue by auduchinok** [fsharp/fslang-suggestions#1273]
F# uses offside rules to power its indentation-based syntax. This is implemented in a component called `LexFilter` that works between lexing and parsing stages. LexFilter keeps track of the indentation blocks and inserts additional `begin`- and `end`-like tokens for the parser.

To help newcomers write F# code, LexFilter has a recovery rule that is used when a token is insufficiently indented. This rule produces a warning, and starts a new indentation block anyway:

<img width="996" alt="Screenshot 2023-05-23 at 12 42 16" src="https://github.com/fsharp/fslang-suggestions/assets/3923587/35d0b529-ee4c-441b-a810-244ee8254d7e">

While the warning is good, the insertion of the block causes many problems for the parsing and analysis when writing code. Consider a case where a user writes a new member:

![Screenshot 2023-04-14 at 11 13 34](https://github.com/fsharp/fslang-suggestions/assets/3923587/6f892347-4846-4db1-9e31-7139d21c09d0)

There are subsequent definitions and expressions, and all of them are parsed and analyzed correctly.
As soon as the user types `=` in, the LexFilter begins to expect an indented body below, doesn't find it, and applies the recovery rule:

![Screenshot 2023-04-14 at 11 13 39](https://github.com/fsharp/fslang-suggestions/assets/3923587/3af112b6-2dd7-4933-9979-34e991ee7b3c)

That makes all the code below parsed incorrectly as the body of the property. Very often it makes the parser to fail to parse the file at all. In the example above, the lenses below the edited place are gone. They will reappear when you type an expression after `=`, making the code move up and down and blink. There's an unexpected error about attributes inside an expression. The `P` type is inferred incorrectly.

The bigger a file gets, the more unexpected parser/analysis errors appear, because neither the parser nor the type checker expect arbitrary code below to be parsed as the indented block. This leads to blinking highlightings, moving code in editor (because of appearing/disappearing lenses/hints), and features like Code Completion or Parameter Info (Signature Help) become significantly less precise, which is crucial when editing code.

There're many cases where it breaks deep inside expressions, like in the binary expression here:

```fsharp
let a =
    1 +

let b = 1
```

**I propose we** rework how wrong indentation is handled. It's possible to change LexFilter, so it doesn't start a new indentation block with the wrong indentation, and to introduce new parser recovery rules, so it would show errors about missing expressions/definitions and continue parsing.

## Pros and Cons 

**The advantages of making this adjustment to F# are**

Doing that will improve parser recovery, which in turn make the analysis significantly more stable and precise. It can improve the editing experience a lot, since there will be much less blinking, code jumping, and features that rely on inferred types will be much more stable.

**The disadvantages of making this adjustment to F# are**

This is a breaking change, and some code that can be compiled today will no longer compile. Such code produces warnings today and should be adjusted to have proper indentation. I expect such code to be extremely rare and mostly produced by generators (so likely only generators will need updates).

## Extra information

**Estimated cost (XS, S, M, L, XL, XXL):** M-L

I've made a small prototype with an additional parser rule for `let` bindings, and it shows that this recovery works as expected, with an additional error produced, see the [gist](https://gist.github.com/auduchinok/aa0ee96259b684e6983cf89dd8a014d5). It's possible to improve the errors text. The main thing is that subsequents definitions are parsed correctly.

Since it's a breaking change, we should only introduce it in a new language version. It's possible to make it into F# 8 now, so the tooling will be able to use the improvements in the fall, when the new version is published. If we don't ship it in F# 8, the closest date when improvements could be used would be ~November 2024.

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [ ] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [ ] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by dsyme** [comment]
The problem is, I've seen beginner code with literally many of these warnings, and the warning turned off.

I started searching for stuff in github that turns of warning 58, here are two examples:

* https://github.com/ServiceStack/dotnet-app/blob/86a7a83d404fb4689af2778c83b98cb97f5cf760/src/Web/FSharpLangInfo.cs#L24
* https://github.com/lvyitian/CoFlows-CE/blob/ccfb36347bf83767c31e60f5b90afa61bb779f80/QuantApp.Engine/QuantApp.Engine.lnx.fsproj#L5

**Comment by dsyme** [comment]
An alternative might be to have a flag `--strict:SOME-NUMBER` that is turned on by default in new project templates.

**Comment by baronfel** [comment]
Slightly tweaked query that only includes fsproj, props, and targets file: https://github.com/search?q=FS0058+%28path%3A*.fsproj+OR+path%3A*.props+OR+path%3A*.targets%29&type=code

There are only 20 occurrences here, and the majority of those are in two projects - Pulumi F# extensions and the CoFlows family of projects already linked. That doesn't seem like an insurmountable hurdle to overcome.

**Comment by auduchinok** [comment]
If we don't enable it by default in existing projects, then it seems there would be much less point in improving it at all. From the tooling standpoint we want a better parser to have the broadest reach possible. We can, for sure, provide a way to opt-out and to revert to the current warnings, if for some reason it's not easy to correct wrong indentation when SDK is released.

**Comment by dsyme** [comment]
I mean I guess an alternative would be to parse all sources twice, and see if the warning occurs anywhere. If not then go ahead with the better parser.

Anyway, it's ultimately up to @vzarytovskii, @baronfelm and others to decide policy on this sort of thing.  I'm OK with the change but it's a breaking change for sure, and it will hit companies where the developers who wrote the code are no longer even at the company.

**Comment by vzarytovskii** [comment]
> I mean I guess an alternative would be to parse all sources twice, and see if the warning occurs anywhere. If not then go ahead with the better parser.
> 
> Anyway, it's ultimately up to @vzarytovskii, @baronfel and others to decide policy on this sort of thing.  I'm OK with the change but it's a breaking change for sure, and it will hit companies where the developers who wrote the code are no longer even at the company. 
> 
> 
> 
> 
> 
> 
> 

Yeah, my fear is that such codd can be in a bunch of  closed codebases, which "just work".

**Comment by baronfel** [comment]
Super valid points. Templates are a great way to kickstart adoption of features like this, and we also have knobs like defaulting the new parser to off for .NET 8, and immediately in 9 preview 1 builds turning the parser on, that way we have the entire 9 preview cycle to find and react to bugs or impacts in the tech.  I pinged Kathleen so we could chat about what our overall strategy for these kinds of changes should be (which fits nicely into a "rules that folks who contribute to the .NET SDK should follow for breaking changes and diagnostics" thing I'm working on).

**Comment by vzarytovskii** [comment]
@auduchinok I wonder how hard it will be to hide new recovery path under a flag?

**Comment by auduchinok** [comment]
> @auduchinok I wonder how hard it will be to hide new recovery path under a flag?

Should be easy, since it's going to be a language version dependent change anyway. LexFilter changes will look at the flag or language version and control this, while the new parser rules just won't be triggered in the most (if not all) cases. The parser rules are going to be just new recovery rules, like the ones I've been adding recently.

**Comment by cartermp** [comment]
FWIW I would be in favor of:

* Enable a bunch of these fixes by default for everyone
* Have a straightforward setting in project files to go back to old behavior (and VS tools checkbox in project settings for that), then document this and include some message about it in the error message

I think we should also weigh the disruption to some against other kinds of disruptions. I have no doubt that we've done many "acceptable" things, especially in tools, that are far more disruptive than something like this.

That doesn't excuse this change from needing to be done responsibly, but if the net benefit is that newcomers to the language can just type code like they expect and it usually "just works" then I think that's big _also_ for these larger, closed-source enterprise codebases where people need to maintain code they didn't write.

**Comment by Tarmil** [comment]
> I think we should really follow this path for this one:
> 
> > ....have a flag `--strict:SOME-FEATURE-NAME` that is turned on by default in new project templates....
> 
> There will 100% be significant enterprise code that emits and suppresses this warning. We can't have installing new versions of the F# compiler or .NET SDK just randomly start giving a hard error for suppressed instances of this warning, or indeed even for unsuppressed instances.

Alternatively, we could consider revising the rule that if a language version is not specified, then the latest is used. And instead, use v7 -- which would be the last version that had this policy -- and show a warning or lint that you should specify a language version.

**Comment by vzarytovskii** [comment]
> > I think we should really follow this path for this one:
> > 
> > > ....have a flag `--strict:SOME-FEATURE-NAME` that is turned on by default in new project templates....
> > 
> > There will 100% be significant enterprise code that emits and suppresses this warning. We can't have installing new versions of the F# compiler or .NET SDK just randomly start giving a hard error for suppressed instances of this warning, or indeed even for unsuppressed instances.
> 
> Alternatively, we could consider revising the rule that if a language version is not specified, then the latest is used. And instead, use v7 -- which would be the last version that had this policy -- and show a warning or lint that you should specify a language version.

We want people to use latest version if its not specified, since it's usually contains a whole bunch of diagnostics, better codegen, new features.

It's better to have a separate flag for this feature.

**Comment by auduchinok** [comment]
> We want people to use latest version if its not specified, since it's usually contains a whole bunch of diagnostics, better codegen, new features.
> It's better to have a separate flag for this feature.

It seems the best way to do this would be to have a flag and to enable/disable it by default based on the language version. This way people using the new SDK will get the changes automatically, while it will still be possible to disable it, or to enable it if an older language version is used explicitly in the project file for some reason.

**Comment by auduchinok** [comment]
@dsyme @vzarytovskii If you agree with this proposal, could it get an approval, please?

If there're no objections to trying to get it into .NET 8, then I'll start working on it.

**Comment by dsyme** [comment]
I'll mark it as approved.  

By approving I mean if the feature applies then this strawman (please let's capture this very precisely):
* make the "offside warning" into a hard error
* make the proposed adjustment to the grammar to improve indentation recovery
* implement this by duplicating the entire parser
* take advantage of this duplication to also completely flush all the non-indentation-aware rules from the copy of the parser
* also flush any other rules that give hard errors on long deprecated features
* put the whole thing in preview so we can perhaps pile in some other changes, since we don't get to do this often
* if langversion < PREVIEW the old parser is used 
* I'd favour using the "immediately on-by-default-at-next-major-upgrade-of-.NET-SDK" mechanism @baronfel mentioned to get this to be the default behaviour.

Now I think of it, duplicating the parser is actually problematic because the token definitions all get duplicated. That would require a change in FsLexYacc. Ugh.  I really feel this needs it though.... @auduchinok Do you know a way that would avoid this duplication while retaining certainty that we haven't mucked with parsing of old code?

**Comment by auduchinok** [comment]
> implement this by duplicating the entire parser

> Now I think of it, duplicating the parser is actually problematic because the token definitions all get duplicated. That would require a change in FsLexYacc. Ugh. I really feel this needs it though.... @auduchinok Do you know a way that would avoid this duplication while retaining certainty that we haven't mucked with parsing of old code?

On the parser side, this change should only require adding/changing recovery rules, like the ones I've been adding recently, not changing normal case ones. It should have no more effect on normal rules than lots of recently added recovery ones.

I think splitting the parser into the old and the new ones can be good in the long term but should probably be done separately, as it's really orthogonal to the indentation recovery I'm proposing. This task is big enough by itself even without this additional refactoring. 🙂

**Comment by kerams** [comment]
Love the idea of completing the sunsetting non-light syntax and ML compatibility.

**Comment by dsyme** [comment]
> On the parser side, this change should only require adding/changing recovery rules, like the ones I've been adding recently, not changing normal case ones. It should have no more effect on normal rules than lots of recently added recovery ones.

But don't those rules apply unconditionally? Making it impossible for that parser to parse like it did previously?

**Comment by auduchinok** [comment]
> But don't those rules apply unconditionally? Making it impossible for that parser to parse like it did previously?

Yes, they should, but we have control over what's inserted by `LexFilter`, and we can make the new recovery rules specific to the LexFilter changes where needed.

**Comment by Tarmil** [comment]
Among the obsolete rules that could be flushed by a parser clean slate, #806 is also a candidate. Infix operator de-indentation was considered a misfeature by everyone in the conversation and it bites people regularly. Don approved adding a warning on it; but while we're making breaking changes to the parser, we could even remove it altogether.

**Comment by auduchinok** [comment]
I've implemented the new rules for the most common cases in https://github.com/dotnet/fsharp/pull/15334, and it improved parsing recovery significantly, fixing many issues with highlighting/folding/lenses during typing, like reported in https://github.com/dotnet/fsharp/issues/7078.

With this change places where code gets deindented act as syncronization points, and it allows producing `begin`/`end`-like tokens for the parser, instead of making it parse everything below unfinished code incorrectly. Since F# syntax is indentation based, using indentation blocks as recovery mechanism seems to be natural and probably the best choice we have. In other words, every place in the language that required additional indentation now allows us to significantly improve things, and every place that didn't may make it more challenging (or impossible) to improve the tooling support.

Here're some examples:
```fsharp
let _ =
  if true then
  () // warning
```
```fsharp
let _ =
  if true then () else

  () // no warning, so it's possible to write early returns without additional indentation
```
```fsharp
let _ =
  match () with
  | _ ->

  () // no warning, a similar reason
```

These cases deliberately allow code to not have additional indentation, and it makes the usage of the language much better in some scenarios. Fortunately, the most of such cases normally appear inside other expressions/declarations, so the outer indentation block still help us have better recovery.

Unfortunately, there're cases where the language design doesn't seem to be consistent, and some cases don't require additional indentation, despite code below often becomes impossible to be parsed (in some expected way or at all):

```fsharp
type T =
member this.Prop = 1 // warning is produced (error with the new rules)
```

```fsharp
type T with
member this.Prop = 1 // no warning
```

Not having indentation here makes it impossible to properly parse things below
```fsharp
type T with
member this.Prop = 1

let x = 1 // error, this is considered part of the type augmentation
```

Subsequently, it doesn't parse things as expected while typing too, even if you're going to have the proper indentation:
```fsharp
type T with

let x = 1 // errors here and below, until a member with proper indentation is typed above
```

There're other similar cases:
```fsharp
type T() =
    do
    () // no warning

    member this.Prop = 1 // error, this is parsed as part of the `do` expression
```

And it breaks for unfinished code too, making user experience much worse:
```fsharp
type T() =
    do

    // errors here and below
    // this is unsuccessfully parsed as part of the `do` expression
    member this.Prop = 1 
```

With all this in mind, I propose that we normalize the indentation requirements where it makes sense:
* for top level declarations indentation should be required, if the code below can't normally be parsed without it in the most of the cases (like in the `type`/`with` and `do` examples above)
* the changed places should start producing warnings if the new indentation rules are disabled
* deliberate undentation is going to still be allowed, because it's an important part of the language that people use

We can discuss the particular cases where it makes sense to change the behaviour, and I'll try to start normalizing the requirements in https://github.com/dotnet/fsharp/pull/15042 and https://github.com/dotnet/fsharp/pull/15102.
